### PR TITLE
remove search endpoint at /api/search

### PR DESCRIPTION
**Pull request by rubichandrap**
moved search functionality to client side
added search by categories
added back to top btn

### CODE REVIEW

File: pages/search.tsx [hunk 1]
```diff
@@ -1,33 +1,50 @@
 import { useDebounce } from '#hooks/index';
-import { ChangeEvent, useEffect, useState } from 'react';
+import React, { ChangeEvent, useEffect, useMemo, useState } from 'react';
```
Comment (aldy505) [inline pages/search.tsx:2]: awas di komen dicha soal `import React` wkwk

File: utils/posts.ts [hunk 2]
```diff
@@ -35,26 +36,30 @@ export const getAllPosts = async (fields: Fields = []) => {
...
-        value.forEach((v) => {
-          if (v.toLowerCase().includes(lowerCasedKeywords)) matchCount += 1;
-        });
+// this function will return all categories available such ['javascript', 'tutorial'] from all posts dynamically
+// but based from my reference which is medium, they're using
+// something like 'people', 'story', 'tags', etc.
+// so rather than returning all categories available
+// I'm using constant string[] from a couple of available meta from each post
+// so this function quite unused
+export const getPostCategories = async (posts?: Partial<PostFields>[]) => {
```
Comment (aldy505) [inline utils/posts.ts:50]: understandable. might use this function later down the road tho👍

File: utils/posts.ts [hunk 3]
```diff
@@ -35,26 +36,30 @@ export const getAllPosts = async (fields: Fields = []) => {
...
+  let _posts = posts ?? (await getAllPosts(['categories']));
```
Comment (aldy505) [inline utils/posts.ts:51]: might as well use `posts ??= await getAllPosts(['categories'])`

File: utils/posts.ts [hunk 4]
```diff
@@ -35,26 +36,30 @@ export const getAllPosts = async (fields: Fields = []) => {
...
+  const categories = new Set();
 
-        return !!matchCount;
+  for (let i = 0; i < _posts.length; i += 1) {
+    let post = _posts[i];
```
Comment (aldy505) [inline utils/posts.ts:55]: prefer const, but I guess this would be fine

**Review by aldy505 — APPROVED** [review]

File: pages/search.tsx (hunk 1 above)
Comment (rubichandrap) [reply pages/search.tsx:2]: ini auto import soalnya, ketika make event bawaan React buat type, such React.MouseEvent atau React.FocusEvent etc, Reactnya kepanggil juga. Kadang yang begini ga di cek dan lupa di apus.

File: utils/_modules.ts [hunk 5]
```diff
@@ -0,0 +1,64 @@
+import type { PostFields } from '#types/post';
+
+export const filterPostsByAvailableFields = (
+  posts: Partial<PostFields>[],
+  lookup: string[] = ['author', 'title', 'categories', 'desc'],
```
Comment (elianiva) [inline utils/_modules.ts:5]: yg ini harus always include `github`, soalnya dipake buat foto authornya
kalo ga diikutin fotonya ga bakal keload soalnya ngarah ke `https://github.com/undefined`

**Review by elianiva — CHANGES_REQUESTED** [review]

File: pages/search.tsx (hunk 1 above)
Comment (elianiva) [reply pages/search.tsx:2]: > awas di komen dicha soal import React wkwk

wkwkwkw tadinya mau gitu, tapi biarin deh, ga penting penting amat

File: utils/_modules.ts (hunk 5 above)
Comment (rubichandrap) [reply utils/_modules.ts]: sip, berarti untuk lookup proses  tetep dari ['author', 'title', 'categories', 'desc'], tapi sebelum di return tambah pake fields sebelumnya

**Review by elianiva — APPROVED** [review]
thanks

image

**Pull request merged**